base/strings/annotated.jl

```diff
@@ -386,3 +386,62 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
 Get all annotations of `chr`.
 """
 annotations(c::AnnotatedChar) = c.annotations
+
+## AnnotatedIOBuffer
+
+struct AnnotatedIOBuffer <: IO
+    io::IOBuffer
+    annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}
+    dict::ImmutableDict{Symbol, Any}
+end
+
+AnnotatedIOBuffer(io::IOBuffer, dict::ImmutableDict = ImmutableDict{Symbol,Any}()) =
+    AnnotatedIOBuffer(io, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}(), dict)
+
+function AnnotatedIOBuffer(io::IOBuffer, (k, v)::Pair{Symbol, Any}, kvs::Pair{Symbol, Any}...)
+    dict = ImmutableDict{Symbol,Any}(k, v)
+    for (k, v) in kvs
+        dict = ImmutableDict{Symbol,Any}(dict, k, v)
+    end
+    AnnotatedIOBuffer(io, dict)
+end
+
+AnnotatedIOBuffer() = AnnotatedIOBuffer(IOBuffer())
+
+function show(io::IO, annio::AnnotatedIOBuffer)
+    show(io, AnnotatedIOBuffer)
+    print(io, '(', annio.io.size, " bytes)")
+end
+
+position(io::AnnotatedIOBuffer) = position(io.io)
+lock(io::AnnotatedIOBuffer) = lock(io.io)
+unlock(io::AnnotatedIOBuffer) = unlock(io.io)
+get(io::AnnotatedIOBuffer, key, default) = key == :color || get(io.dict, key, default)
+displaysize(io::AnnotatedIOBuffer) =
+    if haskey(io.dict, :displaysize) io.dict[:displaysize]::Tuple{Int,Int} else displaysize(io.io) end
```

Comment:
Are you following some sort of documented interface for how to define a fully featured IO type, or is this just a classic julian undocumented set of methods that happen to do the trick? (It's not your responsibility to document how to define a new IO type, but if it is documented somewhere, I'd love to know about it)

Reply:
I looked at the implementation of `IOContext` (since this works similarly), and then this is a case of good ol' copy + paste + tweak (I'm not aware of any documented interface, but maybe I missed something?).